library/src/tests/resources/state_preparation.qs

```diff
@@ -0,0 +1,137 @@
+            fail "Cat state preparation failed.";
+        }
+    }
+
+    operation PrepareComplex(qs : Qubit[]) : Unit is Adj {
+        H(qs[0]);
+        S(qs[0]);
+        H(qs[1]);
+        T(qs[1]);
+    }
+
+    operation TestPrepareComplex(): Unit {
+        use q = Qubit[2];
+        let c00 = ComplexPolar(0.5, 0.0);
+        let c01 = ComplexPolar(0.5, PI()/4.0);
+        let c10 = ComplexPolar(0.5, PI()/2.0);
+        let c11 = ComplexPolar(0.5, 3.0*PI()/4.0);
+        ApproximatelyPreparePureStateCP(0.0, [c00, c01, c10, c11], q);
+        Adjoint PrepareComplex(q);
+        if ( not CheckAllZero(q) ) {
+            fail "Complex state preparation failed.";
+        }
+    }
+
+    operation TestPreparationCompletion(): Unit {
+        let testCases = [
+            // Test positive coefficients
+            [0.773761, 0.633478],
+            [0.183017, 0.406973, 0.604925, 0.659502],
+            [0.0986553, 0.359005, 0.465689, 0.467395, 0.419893, 0.118445, 0.461883, 0.149609],
+            [0.271471, 0.0583654, 0.11639, 0.36112, 0.307383, 0.193371, 0.274151, 0.332542, 0.130172, 0.222546, 0.314879, 0.210704, 0.212429, 0.245518, 0.30666, 0.22773],
+
+            // Test negative coefficients; should give same probabilities as positive coefficients
+            [-0.773761, 0.633478],
+            [0.183017, -0.406973, 0.604925, 0.659502],
+            [0.0986553, -0.359005, 0.465689, -0.467395, 0.419893, 0.118445, -0.461883, 0.149609],
+            [-0.271471, 0.0583654, 0.11639, 0.36112, -0.307383, 0.193371, -0.274151, 0.332542, 0.130172, 0.222546, 0.314879, -0.210704, 0.212429, 0.245518, -0.30666, -0.22773],
+
+            // Test unnormalized coefficients
+            [1.0986553, 0.359005, 0.465689, -0.467395, 0.419893, 0.118445, 0.461883, 0.149609],
+
+            // Test missing coefficients
+            [1.0986553, 0.359005, 0.465689, -0.467395, 0.419893, 0.118445]
+        ];
+
+        for coefficients in testCases {
+            let L = Length(coefficients);
+            let N = Ceiling(Log(IntAsDouble(L))/LogOf2() - 0.001);
+            use q = Qubit[N];
+            PreparePureStateD(coefficients, q);
+            // NOTE: We cannot check that the state is actually correct.
+            if CheckAllZero(q) {
+                fail "Failed to prepare pure state. L={L}, N={N}, coefficients={coefficients}.";
+            }
+            Adjoint PreparePureStateD(coefficients, q);
+            if not CheckAllZero(q) {
+                fail "Failed to prepare and unprepare pure state. L={L}, N={N}, coefficients={coefficients}.";
+            }
+        }
+    }
+
+    operation CheckEndianness(attempts: Int, number: Int, qs: Qubit[]): Unit {
+        let n = Length(qs);
+        for attempt in 1..attempts {
+            mutable c = Repeated(0.0, 2^n);
+            set c w/= number <- 1.0;
+            PreparePureStateD(c, qs);
+            let measured = MeasureInteger(Reversed(qs));
+            if number == measured {
```

Comment:
(If we end up adopting Stefan's suggestion on tests, we can replace iteration and measurement with that approach to validate the state directly)

Reply:
Having stable tests is more favorable than having tests that detect all failures. That is if we have to choose. Like Stefan says it's about time to add a bit of infrastructure so that we don't have to choose.

Reply:
We are inspecting the actual state now.